Explorer/Assets/DCL/SDKComponents/AvatarAttach/Systems/AvatarAttachHandlerSystem.cs

```diff
@@ -0,0 +1,134 @@
+using DCL.SDKComponents.AvatarAttach.Components;
+using DCL.Utilities;
+using ECS.Abstract;
+using ECS.Groups;
+using ECS.LifeCycle;
+using ECS.LifeCycle.Components;
+using ECS.Unity.Transforms.Components;
+using UnityEngine;
+
+namespace DCL.SDKComponents.AvatarAttach.Systems
+{
+    [UpdateInGroup(typeof(SyncedPostRenderingSystemGroup))]
+    [LogCategory(ReportCategory.AVATAR_ATTACH)]
+    [ThrottlingEnabled]
+    public partial class AvatarAttachHandlerSystem : BaseUnityLoopSystem, IFinalizeWorldSystem
+    {
+        private AvatarBase ownPlayerAvatarBase;
+
+        public AvatarAttachHandlerSystem(World world, WorldProxy globalWorld) : base(world)
+        {
+            globalWorld.Query(new QueryDescription().WithAll<PlayerComponent, AvatarBase>(), (ref PlayerComponent playerComponent, ref AvatarBase avatarBase) => { ownPlayerAvatarBase = avatarBase; });
+        }
+
+        protected override void Update(float t)
+        {
+            SetupAvatarAttachQuery(World);
+            UpdateAvatarAttachTransformQuery(World);
+            HandleComponentRemovalQuery(World);
+            HandleEntityDestructionQuery(World);
+        }
+
+        [Query]
+        [None(typeof(AvatarAttachComponent))]
+        private void SetupAvatarAttach(in Entity entity, ref TransformComponent transformComponent, ref PBAvatarAttach pbAvatarAttach)
+        {
+            var component = new AvatarAttachComponent();
+
+            component.anchorPointTransform = GetAnchorPointTransform(pbAvatarAttach.AnchorPointId);
+
+            ApplyAnchorPointTransformValues(transformComponent, component);
+            transformComponent.UpdateCache();
+            World.Set(entity, transformComponent);
+
+            World.Add(entity, component);
+        }
+
+        [Query]
+        private void UpdateAvatarAttachTransform(in Entity entity, ref PBAvatarAttach pbAvatarAttach, ref AvatarAttachComponent avatarAttachComponent, ref TransformComponent transformComponent)
+        {
+            if (pbAvatarAttach.IsDirty)
+            {
+                avatarAttachComponent.anchorPointTransform = GetAnchorPointTransform(pbAvatarAttach.AnchorPointId);
+                World.Set(entity, avatarAttachComponent);
+            }
+
+            if (ApplyAnchorPointTransformValues(transformComponent, avatarAttachComponent))
+            {
+                transformComponent.UpdateCache();
+                World.Set(entity, transformComponent);
+            }
+        }
+
+        [Query]
+        [All(typeof(AvatarAttachComponent))]
+        [None(typeof(PBAvatarAttach), typeof(DeleteEntityIntention))]
+        private void HandleComponentRemoval(in Entity entity)
+        {
+            World.Remove<AvatarAttachComponent>(entity);
+        }
+
+        [Query]
+        [All(typeof(DeleteEntityIntention), typeof(AvatarAttachComponent))]
+        private void HandleEntityDestruction(in Entity entity)
+        {
+            World.Remove<AvatarAttachComponent>(entity);
```

Comment:
Same here; moreover there is a critical vulnerability:
- You pass `Entity` by `readonly ref` (which is `in`)
- When you make a structural change the memory layout is changed so `entity` will no longer point to the same entity
- You could replace it with `World.Remove<AvatarAttachComponent, PBAvatarAttach>`

But it's much more perfomant to use a bulk request: `World.Remove<AvatarAttachComponent, PBAvatarAttach>(YOUR_QUERY);` by calling it from `Update` directly

Reply:
Interesting, makes sense. Done.